### PR TITLE
[Hetero] Heterograph C++ implementation; Bipartite and Python wrapper

### PR DESCRIPTION
**Pull request by jermainewang**
## Description
The PR implements bipartite graph and heterograph C++ data structure. As per our discussion, the bipartite graph is implemented similarly to immutable graph which contains on-demand-materialized CSR/COO structure. Heterograph is implemented as a list of bipartite graphs, each for one edge type. On python side, currently only minimal wrapper class is implemented.

Side note:
I feel we should hide the COO/CSR implementation from header file as what is done in [bipartite.h](https://github.com/dmlc/dgl/blob/682a3a9510667a29fbbd3ec35aeca582e8685a5a/src/graph/bipartite.h). It makes the header file much cleaner.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: src/graph/heterograph.cc [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+      const dgl_type_t src_vtype = pair.first;
+      const dgl_type_t dst_vtype = pair.second;
+      subrels[etype] = Bipartite::CreateFromCOO(
+        ret.induced_vertices[src_vtype]->shape[0],
+        ret.induced_vertices[dst_vtype]->shape[0],
+        subedges[etype].src,
+        subedges[etype].dst);
+    }
+    ret.graph = HeteroGraphPtr(new HeteroGraph(meta_graph_, subrels));
+  }
```
Comment (BarclayII) [inline src/graph/heterograph.cc]: Divide into two functions?

File: include/dgl/base_heterograph.h [hunk 2]
```diff
@@ -102,48 +107,13 @@ class HeteroGraphInterface {
...
-    dgl_id_t* rst_data = static_cast<dgl_id_t*>(rst->data);
-    for (int64_t i = 0; i < len; ++i) {
-      rst_data[i] = HasVertex(vtype, vid_data[i])? 1 : 0;
-    }
-    return rst;
-  }
+  virtual BoolArray HasVertices(dgl_type_t vtype, IdArray vids) const = 0;
 
   /*! \return true if the given edge is in the graph.*/
   virtual bool HasEdgeBetween(dgl_type_t etype, dgl_id_t src, dgl_id_t dst) const = 0;
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply include/dgl/base_heterograph.h]: Good question. What I'm thinking about is the graph class is type-agnostic (i.e, use NDArray). For scalar type, we could use int64 but the implementation might cast it to int32 if needed.

File: python/dgl/graph_index.py [hunk 3]
```diff
@@ -1235,4 +1253,737 @@ def create_graph_index(graph_data, multigraph, readonly):
...
+        utils.Index
+            The edge ids.
+        """
+        u_array = u.todgltensor()
+        v_array = v.todgltensor()
+        edge_array = _CAPI_DGLHeteroEdgeIds(self, int(etype), u_array, v_array)
+
+        src = utils.toindex(edge_array(0))
+        dst = utils.toindex(edge_array(1))
+        eid = utils.toindex(edge_array(2))
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/graph_index.py]: I think so. For consistency, we should try use object, but this practice is also fine.

File: src/graph/bipartite.cc [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+  }
+
+  HeteroGraphPtr GetRelationGraph(dgl_type_t etype) const override {
+    LOG(FATAL) << "The method shouldn't be called for Bipartite graph. "
+      << "The relation graph is simply this graph itself.";
+    return {};
+  }
+
+  void AddVertices(dgl_type_t vtype, uint64_t num_vertices) override {
+    LOG(FATAL) << "Bipartite graph is not mutable.";
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply src/graph/bipartite.cc]: I'm thinking about getting rid of adjlist graph in the future due to (1) maintaining two types are costly; (2) we could use COO to implement mutable class by array concatenating and make CSR a thread-safe volatile member. In fact, the removal operation is already done using subgraph. Also, using all ndarray data structure makes it easier to ship the graph to GPU.

File: src/graph/bipartite.cc [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+
+  BoolArray HasVertices(dgl_type_t vtype, IdArray vids) const override {
+    LOG(FATAL) << "Not enabled for COO graph";
+    return {};
+  }
+
+  bool HasEdgeBetween(dgl_type_t etype, dgl_id_t src, dgl_id_t dst) const override {
+    CHECK(HasVertex(0, src)) << "Invalid src vertex id: " << src;
+    CHECK(HasVertex(1, dst)) << "Invalid dst vertex id: " << dst;
+    return aten::CSRIsNonZero(adj_, src, dst);
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply src/graph/bipartite.cc]: That's fine. Non-zero is quite well-accepted in terms of sparse matrix.

File: src/graph/heterograph.cc (hunk 1 above)
Comment (jermainewang) [reply src/graph/heterograph.cc]: Done.

File: src/graph/bipartite.cc [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+#include <dgl/immutable_graph.h>
+
+#include "./bipartite.h"
+#include "../c_api_common.h"
+
+namespace dgl {
+namespace {
+inline GraphPtr CreateBipartiteMetaGraph() {
+  static IdArray row = aten::VecToIdArray(std::vector<int64_t>({0}));
+  static IdArray col = aten::VecToIdArray(std::vector<int64_t>({1}));
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply src/graph/bipartite.cc]: Done.

**Comment by jermainewang** [comment]
@BarclayII strange WIN DLL linking error. Could you have a look?

File: src/graph/bipartite.cc [hunk 7]
```diff
@@ -0,0 +1,824 @@
...
+#include <dgl/lazy.h>
+#include <dgl/immutable_graph.h>
+
+#include "./bipartite.h"
+#include "../c_api_common.h"
+
+namespace dgl {
+namespace {
+inline GraphPtr CreateBipartiteMetaGraph() {
+  IdArray row = aten::VecToIdArray(std::vector<int64_t>({Bipartite::kSrcVType}));
```
Comment (BarclayII) [inline src/graph/bipartite.cc]: Changing `int64_t` to `uint64_t` or `dgl_type_t` gets rid of the bug.

Longer story:
When I separated `std::vector<int64_t>({Bipartite::kSrcVType})` into an assignment statement:
```c++
std::vector<int64_t> row_vec = {Bipartite::kSrcVType};
```
MSVC throws a compilation error saying it requires a narrowing conversion.  However, when embedding this faulty assignment statement into an expression like you did here, MSVC somehow ignores it and outputs a vector with size 0 (i.e. generating wrong code).

Issue verified with the following code:
```c++
int size(const std::vector<int64_t> &v)
{
	return v.size();
}

int main()
{
	std::vector<int64_t> v({ (uint64_t)0 });
	std::cout << size(v) << std::endl;  // prints 0
	return 0;
}
```

I will find some time to search or raise the issue in MSDN.  Meanwhile, please just change `int64_t` to `uint64_t` or `dgl_id_t` here.

**Pull request merged**